### ISSUE TITLE
Issue with pairing

### ISSUE THREAD

**Issue by johnappletree** [ejurgensen/forked-daapd#587]
Hi There, 

I am running forked-daapd 26.2 in a jail on freebsd 11.2. I used the install script for freebsd found in the repository. I am trying to connect my iPhone with the server. however the webinterface does not seem to recognize my iOS device. When I use the suggested avachi browse command, my iphone request is listed.  (similar like the example)

+ epair0b IPv4 xxxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxxx          _touch-remote._tcp   local
= epair0b IPv4 xxxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxxx          _touch-remote._tcp   local
   hostname = [iPhone-John.local]
   address = [192.168.1.25]
   port = [52361]
   txt = ["DvTy=iPhone" "RemN=iTunes Remote" "txtvers=1" "Pair=FF26FXXXXXXX49" "RemV=10000" "DvNm=iPhone John"]

I also checked the log file but nothing related is found there.

Regards,

John

**Comment by ejurgensen** [comment]
Try setting the log level to info in the config. If avahi (in the jail) is detecting the device then so should forked-daapd, at it should be logged. If there is still nothing in the log then I'm not sure what is going on...

**Comment by johnappletree** [comment]
Thanks for pointing me in the right direction. Log says this:
[ WARN] mdns: Ignoring announcement from iPhone-John.local, address 192.168.1.25 is not connectable
[ WARN] mdns: Ignoring announcement from iPhone-John.local, address 192.168.1.25 is not connectable

What could be the reason for this?

**Comment by ejurgensen** [comment]
Don't know, try to make the connection with telnet, so from the jail run "telnet 192.168.1.25 xxx" where xxx is the announced port. I think the port changes, so grab it with avahi-browse like above. It will show if it is a network problem or if forked-daapd is the issue.

If you set the log level further down to debug you might see more info, but I'm guessing it will just say that there was a timeout trying to make the connection.

**Comment by ejurgensen** [comment]
You could try increasing the timeout, don't know if forked-daapd is just too impatient. Since you are building yourself it should be relatively easy, just increase MDNS_CONNECT_TEST_TIMEOUT in src/mdns_avahi.c, rebuild and try again. Let me know if that does anything.

**Comment by johnappletree** [comment]
Same problem when i increase the timeout:

```
[2018-09-09 11:38:04] [DEBUG]     mdns: Connection test to 192.168.1.25:55496 timed out (limit is 10 seconds)
[2018-09-09 11:38:04] [ WARN]     mdns: Ignoring announcement from iPhone-John.local, address 192.168.1.25 is not connectable
[2018-09-09 11:38:04] [DEBUG]     mdns: Avahi Record Browser (iPhone-John.local, proto 0): no more results (CACHE_EXHAUSTED)
```

**Comment by johnappletree** [comment]
Nice That worked! Shall I send a pull request? Or do you want to keep the connection test in the script?

**Comment by ejurgensen** [comment]
There is a reason it's there, so just removing it will lead to other problems. I will look into why it isn't working as intended. I might come up with a fix that needs testing, hope you can help with that since I can't reproduce the issue myself.

**Comment by johnappletree** [comment]
Sure, i will do!

**Comment by cosinekitty** [comment]
I doubt this is the problem here, but I see a possible missed error check in the code.  In `mdns_avahi.c`, the function `address_check()` calls `avahi_address_snprint()` without checking its return value:

```
avahi_address_snprint(address, sizeof(address), addr);
```

Looking at the [source  for `avahi_address_snprint()`](https://github.com/lathiat/avahi/blob/master/avahi-common/address.c), it can return NULL if there is an error. It looks like unpredictable things could happen, including possible use of uninitialized memory in the local variable `address[]`, if an error did occur.

**Comment by ejurgensen** [comment]
@johnappletree I noticed that your log shows that forked-daapd is actually not timing out, but actually failing immediately. Look at the timestamps in the below:
```
[2018-09-08 21:12:32] [DEBUG]     mdns: Avahi Resolver: resolved service 'xxxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxxx' type '_touch-remote._tcp' proto 0, host iPhone-John.local, address 192.168.1.25
[2018-09-08 21:12:32] [DEBUG]     mdns: Connection test to 192.168.1.25:53838 timed out (limit is 2 seconds)
```
So I looked into why that might be, and Stackoverflow tells me that BSD's connect() is more likely to succeed immediately than Linux's, in which case there is really no need to poll for the connection to become writable. I still don't understand why the poll wouldn't just tell me that the connection is writable instead of giving me a false timeout, but never mind that...

The gist of it is that now added commit 2c1f2f1b1529077a9959e6e19b0c1875c27a670b to master (it should be safe) and I think it deals with BSD's immediate connection success, I hope you will pull it, rebuild and test if it now works.

@cosinekitty thanks for pointing to the missing checks, I also added those
